File: plugins/scripts/http.lua
-- $crypt HTTP/1.x Dissector Plugin (Lua)
-- Parses HTTP request and response headers with content analysis.

plugin = {
    name = "HTTP",
    version = "0.2.0",
    author = "$crypt Collective",
    description = "HTTP/1.x request/response dissector with header analysis",
}

-- Common HTTP ports
local http_ports = {
    [80] = true, [8080] = true, [8443] = true,
    [8000] = true, [8888] = true, [3000] = true,
}

-- Check if data looks like HTTP
function can_dissect(data_len, src_port, dst_port)
    if http_ports[dst_port] or http_ports[src_port] then
        return "high"
    end
    return "none"
end

-- Well-known HTTP methods
local http_methods = {
    GET = true, POST = true, PUT = true, DELETE = true, PATCH = true,
    HEAD = true, OPTIONS = true, TRACE = true, CONNECT = true,
}

-- Categorize status code
local function status_category(code)
    local n = tonumber(code) or 0
    if n >= 100 and n < 200 then return "Informational"
    elseif n >= 200 and n < 300 then return "Success"
    elseif n >= 300 and n < 400 then return "Redirection"
    elseif n >= 400 and n < 500 then return "Client Error"
    elseif n >= 500 and n < 600 then return "Server Error"
    else return "Unknown" end
end

-- Detect content type category for summary
local function content_category(ct)
    if not ct then return nil end
    ct = ct:lower()
    if ct:find("text/html") then return "HTML"
    elseif ct:find("application/json") then return "JSON"
    elseif ct:find("application/xml") or ct:find("text/xml") then return "XML"
    elseif ct:find("text/plain") then return "Text"
    elseif ct:find("text/css") then return "CSS"
    elseif ct:find("javascript") then return "JavaScript"
    elseif ct:find("image/") then return "Image"
    elseif ct:find("application/octet") then return "Binary"
    elseif ct:find("multipart/form") then return "Form Data"
    else return nil end
end

-- Dissect HTTP data
function dissect(data, src_port, dst_port)
    local text = tostring(data)
    local fields = {}
    local summary = ""
    local header_end = text:find("\r\n\r\n")
    local header_len = header_end and (header_end + 3) or #text

    -- Headers lookup (lowercase keys for easy access)
    local headers = {}
    local header_count = 0

    -- Check if it's a request or response
    local method, uri, version = text:match("^(%u+)%s+(%S+)%s+(HTTP/%d%.%d)")
    if method and http_methods[method] then
        -- HTTP Request
        table.insert(fields, {
            name = "Method",
            value = method,
            offset = 0,
            len = #method,
        })
        table.insert(fields, {
            name = "URI",
            value = uri,
            offset = #method + 1,
            len = #uri,
        })
        table.insert(fields, {
            name = "Version",
            value = version,
            offset = #method + 1 + #uri + 1,
            len = #version,
        })

        -- Parse headers
        for name, value in text:gmatch("([%w%-]+):%s*(.-)\r\n") do
            headers[name:lower()] = value
            header_count = header_count + 1
            table.insert(fields, { name = name, value = value, offset = 0, len = 0 })
        end

        -- Add host to summary if present
        local host = headers["host"] or ""
        summary = string.format("%s %s %s", method, uri, version)
        if host ~= "" then
            summary = string.format("%s %s%s %s", method, host, uri, version)
        end
    else
        local resp_version, status_code, reason = text:match("^(HTTP/%d%.%d)%s+(%d+)%s+(.-)%\r")
        if resp_version then
            -- HTTP Response
            local cat = status_category(status_code)
            table.insert(fields, {
                name = "Version",
                value = resp_version,
                offset = 0,
                len = #resp_version,
            })
            table.insert(fields, {
                name = "Status Code",
                value = status_code,
                offset = #resp_version + 1,
                len = #status_code,
            })
            table.insert(fields, {
                name = "Reason",
                value = reason,
                offset = #resp_version + 1 + #status_code + 1,
                len = #reason,
            })
            table.insert(fields, {
                name = "Category",
                value = cat,
                offset = 0,
                len = 0,
            })

            -- Parse headers
            for name, value in text:gmatch("([%w%-]+):%s*(.-)\r\n") do
                headers[name:lower()] = value
                header_count = header_count + 1
                table.insert(fields, { name = name, value = value, offset = 0, len = 0 })
            end

            summary = string.format("%s %s %s", resp_version, status_code, reason)

            -- Annotate content type
            local ct = headers["content-type"]
            local ct_cat = content_category(ct)
            if ct_cat then
                summary = summary .. " [" .. ct_cat .. "]"
            end

            -- Content-Length
            local cl = headers["content-length"]
            if cl then
                table.insert(fields, { name = "Body Length", value = cl .. " bytes", offset = 0, len = 0 })
            end
        else
            -- Could be continuation data or partial
            summary = "HTTP data (continuation)"
            -- Still try to parse as headers
            for name, value in text:gmatch("([%w%-]+):%s*(.-)\r\n") do
                header_count = header_count + 1
                table.insert(fields, { name = name, value = value, offset = 0, len = 0 })
            end
        end
    end

    table.insert(fields, 1, {
        name = "Header Count",
        value = tostring(header_count),
        offset = 0,
        len = 0,
    })

    return {
        protocol = "HTTP",
        summary = summary,
        header_len = header_len,
        fields = fields,
        next_protocol = nil,
    }
end

File: crates/sc-protocol/src/builtins/http.rs
use crate::dissector::*;
use crate::model::*;
use sc_core::Protocol;

/// Native HTTP/1.x dissector.
pub struct HttpDissector;

fn is_http_method(s: &[u8]) -> bool {
    matches!(
        s,
        b"GET"
            | b"POST"
            | b"PUT"
            | b"DELETE"
            | b"HEAD"
            | b"OPTIONS"
            | b"PATCH"
            | b"CONNECT"
            | b"TRACE"
    )
}

fn status_category(code: u16) -> &'static str {
    match code {
        100..=199 => "Informational",
        200..=299 => "Success",
        300..=399 => "Redirection",
        400..=499 => "Client Error",
        500..=599 => "Server Error",
        _ => "Unknown",
    }
}

fn content_type_label(ct: &str) -> &'static str {
    let ct_lower = ct.to_ascii_lowercase();
    if ct_lower.contains("json") {
        "JSON"
    } else if ct_lower.contains("html") {
        "HTML"
    } else if ct_lower.contains("xml") {
        "XML"
    } else if ct_lower.contains("javascript") {
        "JavaScript"
    } else if ct_lower.contains("css") {
        "CSS"
    } else if ct_lower.contains("image") {
        "Image"
    } else if ct_lower.contains("octet-stream") {
        "Binary"
    } else if ct_lower.contains("text") {
        "Text"
    } else {
        "Other"
    }
}

/// Try to find end of HTTP headers (double CRLF).
fn find_header_end(data: &[u8]) -> Option<usize> {
    data.windows(4)
        .position(|w| w == b"\r\n\r\n")
        .map(|p| p + 4)
}

/// Parse header lines from header block.
fn parse_headers(header_block: &[u8]) -> Vec<(&str, &str)> {
    let text = match std::str::from_utf8(header_block) {
        Ok(t) => t,
        Err(_) => return Vec::new(),
    };
    let mut headers = Vec::new();
    for line in text.split("\r\n") {
        if let Some((name, value)) = line.split_once(':') {
            headers.push((name.trim(), value.trim()));
        }
    }
    headers
}

impl Dissector for HttpDissector {
    fn id(&self) -> DissectorId {
        DissectorId("http".into())
    }

    fn name(&self) -> &str {
        "HTTP"
    }

    fn can_dissect(&self, data: &[u8], context: &DissectionContext) -> Confidence {
        if context.next_protocol_hint == Some(Protocol::Http) && data.len() >= 16 {
            return Confidence::Exact;
        }
        // Port heuristic
        let http_port = matches!(
            context.src_port,
            Some(80) | Some(8080) | Some(8000) | Some(8888) | Some(3000)
        ) || matches!(
            context.dst_port,
            Some(80) | Some(8080) | Some(8000) | Some(8888) | Some(3000)
        );

        if !http_port || data.len() < 16 {
            return Confidence::None;
        }

        // Quick check: starts with an HTTP method or "HTTP/"
        if data.starts_with(b"HTTP/") {
            return Confidence::High;
        }
        // Check if first word (up to space) is a known method
        if let Some(sp) = data.iter().position(|&b| b == b' ') {
            if sp <= 7 && is_http_method(&data[..sp]) {
                return Confidence::High;
            }
        }
        Confidence::Low
    }

    fn dissect<'a>(
        &self,
        data: &'a [u8],
        _context: &mut DissectionContext,
    ) -> sc_core::Result<DissectedLayer<'a>> {
        let header_end = find_header_end(data).unwrap_or(data.len());
        let header_block = &data[..header_end];
        let text = std::str::from_utf8(header_block).map_err(|_| sc_core::ShadowError::Parse {
            message: "HTTP: invalid UTF-8 in headers".into(),
        })?;

        // First line
        let first_line_end = text.find("\r\n").unwrap_or(text.len());
        let first_line = &text[..first_line_end];

        let mut fields = vec![Field {
            name: "Request/Status Line".into(),
            display_value: first_line.to_string(),
            byte_range: 0..first_line_end,
            field_type: FieldType::String,
        }];

        let summary = if first_line.starts_with("HTTP/") {
            // Response: HTTP/1.1 200 OK
            let parts: Vec<&str> = first_line.splitn(3, ' ').collect();
            let version = parts.first().unwrap_or(&"HTTP/?");
            let status_code: u16 = parts.get(1).and_then(|s| s.parse().ok()).unwrap_or(0);
            let reason = parts.get(2).unwrap_or(&"");

            fields.push(Field {
                name: "Version".into(),
                display_value: version.to_string(),
                byte_range: 0..version.len(),
                field_type: FieldType::String,
            });
            fields.push(Field {
                name: "Status Code".into(),
                display_value: format!("{status_code} ({})", status_category(status_code)),
                byte_range: 0..first_line_end,
                field_type: FieldType::UInt16,
            });

            format!("HTTP {version} {status_code} {reason}")
        } else {
            // Request: GET /path HTTP/1.1
            let parts: Vec<&str> = first_line.splitn(3, ' ').collect();
            let method = parts.first().unwrap_or(&"?");
            let uri = parts.get(1).unwrap_or(&"/");
            let version = parts.get(2).unwrap_or(&"HTTP/?");

            fields.push(Field {
                name: "Method".into(),
                display_value: method.to_string(),
                byte_range: 0..method.len(),
                field_type: FieldType::String,
            });
            fields.push(Field {
                name: "URI".into(),
                display_value: uri.to_string(),
                byte_range: 0..first_line_end,
                field_type: FieldType::String,
            });
            fields.push(Field {
                name: "Version".into(),
                display_value: version.to_string(),
                byte_range: 0..first_line_end,
                field_type: FieldType::String,
            });

            format!("HTTP {method} {uri} {version}")
        };

        // Parse headers
        let header_start = first_line_end + 2; // skip \r\n
        if header_start < header_end {
            let headers = parse_headers(&data[header_start..header_end - 2]); // trim trailing \r\n
            let header_count = headers.len();

            for (name, value) in &headers {
                let display = format!("{name}: {value}");
                fields.push(Field {
                    name: format!("Header: {name}"),
                    display_value: display,
                    byte_range: header_start..header_end,
                    field_type: FieldType::String,
                });
            }

            // Extract useful metadata
            if let Some((_, ct)) = headers
                .iter()
                .find(|(n, _)| n.eq_ignore_ascii_case("Content-Type"))
            {
                fields.push(Field {
                    name: "Content-Type Category".into(),
                    display_value: content_type_label(ct).to_string(),
                    byte_range: header_start..header_end,
                    field_type: FieldType::String,
                });
            }

            if let Some((_, cl)) = headers
                .iter()
                .find(|(n, _)| n.eq_ignore_ascii_case("Content-Length"))
            {
                fields.push(Field {
                    name: "Content-Length".into(),
                    display_value: cl.to_string(),
                    byte_range: header_start..header_end,
                    field_type: FieldType::String,
                });
            }

            fields.push(Field {
                name: "Header Count".into(),
                display_value: format!("{header_count}"),
                byte_range: header_start..header_end,
                field_type: FieldType::UInt16,
            });
        }

        let body = &data[header_end..];

        let node = ProtocolNode {
            protocol: "HTTP".into(),
            byte_range: 0..header_end,
            fields,
            summary,
        };

        Ok(DissectedLayer {
            node,
            remaining: body,
            next_protocol: None,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_http_request() {
        let data =
            b"GET /index.html HTTP/1.1\r\nHost: example.com\r\nAccept: text/html\r\n\r\nbody";

        let mut ctx = DissectionContext::new(16);
        ctx.dst_port = Some(80);

        let conf = HttpDissector.can_dissect(data, &ctx);
        assert!(conf >= Confidence::High);

        let layer = HttpDissector.dissect(data, &mut ctx).unwrap();
        assert_eq!(layer.node.protocol, "HTTP");
        assert!(layer.node.summary.contains("GET"));
        assert!(layer.node.summary.contains("/index.html"));
        assert_eq!(layer.remaining, b"body");

        // Check Host header
        let host = layer.node.fields.iter().find(|f| f.name == "Header: Host");
        assert!(host.is_some());
    }

    #[test]
    fn test_http_response() {
        let data = b"HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nContent-Length: 13\r\n\r\n{\"status\":\"ok\"}";

        let mut ctx = DissectionContext::new(16);
        ctx.src_port = Some(80);

        let layer = HttpDissector.dissect(data, &mut ctx).unwrap();
        assert!(layer.node.summary.contains("200"));
        assert!(layer.node.summary.contains("OK"));

        let ct_cat = layer
            .node
            .fields
            .iter()
            .find(|f| f.name == "Content-Type Category");
        assert!(ct_cat.is_some());
        assert_eq!(ct_cat.unwrap().display_value, "JSON");
    }

    #[test]
    fn test_http_no_match_wrong_port() {
        let data = b"GET /foo HTTP/1.1\r\nHost: x\r\n\r\n";
        let ctx = DissectionContext::new(16);
        let conf = HttpDissector.can_dissect(data, &ctx);
        assert_eq!(conf, Confidence::None);
    }
}
